src/Symfony/Component/Translation/CatalogueBag/FallbackCatalogueBag.php

```diff
@@ -0,0 +1,123 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Translation\CatalogueBag;
+
+use Symfony\Component\Translation\MessageCatalogueInterface;
+use Symfony\Component\Translation\MessageCatalogue;
+use Symfony\Component\Translation\Translator;
+
+/**
+ * Merges the fallback catalogues into the loaded one.
+ *
+ * @author Abdellatif Ait boudad <a.aitboudad@gmail.com>
+ */
+class FallbackCatalogueBag implements CatalogueBagInterface
+{
+    /**
+     * @var MessageCatalogueInterface[]
+     */
+    private $catalogues = array();
+
+    /**
+     * @var CatalogueBagInterface
+     */
+    private $catalogueBag;
+
+    /**
+     * @var array
+     */
+    private $locales;
+
+    /**
+     * @param CatalogueBagInterface $catalogueBag The catalogue bag to use for loading the catalogue.
+     * @param array                 $locales      The fallback locales.
+     */
+    public function __construct(CatalogueBagInterface $catalogueBag, $locales = array())
+    {
+        $this->setFallbackLocales($locales);
+        $this->catalogueBag = $catalogueBag;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getCatalogue($locale)
+    {
+        if (isset($this->catalogues[$locale])) {
+            return $this->catalogues[$locale];
+        }
+
+        $catalogue = $this->catalogueBag->getCatalogue($locale);
+        $this->loadFallbackCatalogues($catalogue);
```

Comment:
This fetches a catalogue from the CatalogueBag and modifies its fallback catalogue chain. As the CatalogueBag internally caches the catalogue instances, now it suddenly holds a catalogue with a fallback chain. Also, clients who retrieved a catalogue from the CatalogueBag earlier may now be surprised to find that it now has fallback catalogues.